dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,582 @@
+  dash::Array<difference_type> g_nle_all(nunits * nunits, team);
+
+  std::vector<difference_type> l_target_count(nunits + 1);
+
+  // Starting offsets of all units
+  std::vector<difference_type> acc_unit_count(nunits + 1);
+  // initial counts
+  dash::team_unit_t       unit{0};
+  const dash::team_unit_t last{static_cast<dart_unit_t>(nunits)};
+
+  acc_unit_count[0] = 0;
+
+  for (; unit < last; ++unit) {
+    auto const u_size        = pattern.local_size(unit);
+    acc_unit_count[unit + 1] = u_size + acc_unit_count[unit];
+  };
+
+  auto const min = static_cast<value_type>(g_min.get());
+  auto const max = static_cast<value_type>(g_max.get());
+
+  auto const                          nboundaries = nunits - 1;
+  std::vector<value_type>             partitions(nboundaries, 0);
+  detail::PartitionBorder<value_type> p_borders(nboundaries, min, max);
+
+  DASH_SORT_LOG_TRACE_RANGE("locally sorted array", lbegin, lend);
+
+  bool done = false;
+
+  do {
+    detail::psort_calculate_boundaries(p_borders, partitions);
+
+    auto const histograms =
+        detail::psort_local_histogram<value_type, difference_type>(
+            partitions, lbegin, lend);
+
+    auto const& l_nlt = histograms.first;
+    auto const& l_nle = histograms.second;
+
+    // detail::psort_global_histogram(l_nlt, l_nle, g_nlt, g_nle);
+    detail::psort_global_histogram(l_nlt, l_nle, g_nlt_nle);
+
+    done = detail::psort_validate_partitions(
+        p_borders, partitions, acc_unit_count, g_nlt_nle);
+
+    // This swap eliminates a barrier as, otherwise, some units may be one
+    // iteration ahead and modify shared data while the others are still not
+    // done
+    if (!done) {
+      std::swap(g_nlt_nle, g_nlt_nle_buf);
+    }
+
+  } while (!done);
+
+  auto const histograms =
+      detail::psort_local_histogram<value_type, difference_type>(
+          partitions, lbegin, lend);
+
+  /* How many elements are less than P
+   * or less than equals P */
+  auto const& l_nlt = histograms.first;
+  auto const& l_nle = histograms.second;
+
+  DASH_ASSERT_EQ(
+      histograms.first.size(), histograms.second.size(),
+      "length of histogram arrays does not match");
+
+  DASH_SORT_LOG_TRACE_RANGE(
+      "final partition borders", partitions.begin(), partitions.end());
+  DASH_SORT_LOG_TRACE_RANGE(
+      "final histograms: l_nlt", l_nlt.begin(), l_nlt.end());
+  DASH_SORT_LOG_TRACE_RANGE(
+      "final histograms: l_nle", l_nle.begin(), l_nle.end());
+
+  /*
+   * Transpose (Shuffle) the final histograms to communicate
+   * the partition distribution
+   */
+  for (std::size_t idx = 1; idx < l_nlt.size(); ++idx) {
+    auto const transposed_unit = idx - 1;
+    DASH_ASSERT_EQ(
+        g_nlt_all.pattern().local_size(
+            static_cast<dash::team_unit_t>(transposed_unit)),
+        l_nlt.size() - 1, "Invalid Array length");
+    auto const offset = transposed_unit * nunits + myid.id;
+
+    g_nlt_all.async[offset] = l_nlt[idx];
```

Comment:
Use `g_nlt_all.async[offset].put(&l_nlt[idx])` to avoid local flush for temporaries (applies to several places throughout)